src/compas_timber/connections/t_butt.py

```diff
@@ -44,19 +44,43 @@ def __data__(self):
         data["mill_depth"] = self.mill_depth
         return data
 
-    def __init__(self, main_beam=None, cross_beam=None, mill_depth=None, **kwargs):
+    def __init__(self, main_beam=None, cross_beam=None, mill_depth=None, fastener=None, **kwargs):
         super(TButtJoint, self).__init__(**kwargs)
         self.main_beam = main_beam
         self.cross_beam = cross_beam
         self.main_beam_guid = kwargs.get("main_beam_guid", None) or str(main_beam.guid)
         self.cross_beam_guid = kwargs.get("cross_beam_guid", None) or str(cross_beam.guid)
         self.mill_depth = mill_depth
         self.features = []
+        if isinstance(fastener, type):
```

Comment:
so if I understand right, there are two mutually exclusive scenarios here:
1. a fastener instance is passed to `__init__`
2. a fastener class is passed as `fastener` 

it is a bad practice to allow passing different types as the same argument as it's confusing to read and hard to test and debug.

something like this would be preferable and maybe would achieve the same goal:
```python
self.base_fastener = None
if fastener is None:
    self.base_fastener = PlateFastener.default_t()
```

Another issue is that there's another scenario, de-serialization, where the fastener instance is already part in the model but not available yet in `__init__`, therefore it's necessary to save its guid and restore it in `restore_elements_from_guids`.

Reply:
I wanted to cover the case where there is no fastener added to the joint. I see two options:

- add a second argument to the Joint __init__() called `add_fastener` which takes a boolean. Assuming that we can add fasteners to almost any joint, this seems unwieldy.
-  pass an instance of the `PlateFastener` and change its dimensions in the `Joint` def. I can do this, but it would essentially mean creating a completely new fastener.
- We could also consider a new `FastenerDefinition` class which becomes the `joint.base_fastener`